testgen/src/read.rs

```diff
@@ -0,0 +1,214 @@
+    let name = path
+        .file_name()
+        .ok_or_else(|| eyre!(format!("test suite with no name found: {}", path.display())))?
+        .to_str()
+        .ok_or_else(|| eyre!(format!("non-UTF-8 suite name found: {}", path.display())))?;
+
+    let mut suites = Vec::new();
+    let mut tests = Vec::new();
+
+    // TODO: iterate in parallel
+    for entry in path.read_dir().wrap_err("could not retrieve entry")? {
+        let entry = entry?;
+        let filetype = entry.file_type().wrap_err("could not retrieve file type")?;
+
+        if filetype.is_dir() {
+            suites.push(read_suite(entry.path().as_path()).wrap_err_with(|| {
+                let path = entry.path();
+                let suite = path.display();
+                format!("error reading sub-suite {suite}")
+            })?);
+            continue;
+        }
+
+        let path = entry.path();
+
+        if path.extension() != Some(OsStr::new("js")) {
+            // Ignore files that aren't executable.
+            continue;
+        }
+
+        if path
+            .file_stem()
+            .is_some_and(|stem| stem.as_encoded_bytes().ends_with(b"FIXTURE"))
+        {
+            // Ignore files that are fixtures.
+            continue;
+        }
+
+        let test = read_test(&path).wrap_err_with(|| {
+            let path = entry.path();
+            let suite = path.display();
+            format!("error reading test {suite}")
+        })?;
+
+        tests.push(test);
+    }
+
+    Ok(TestSuite {
+        name: name.into(),
+        path: Box::from(path),
+        suites: suites.into_boxed_slice(),
+        tests: tests.into_boxed_slice(),
+    })
+}
+
+/// Reads information about a given test case.
+pub(super) fn read_test(path: &Path) -> Result<Test> {
+    let name = path
+        .file_stem()
+        .ok_or_else(|| eyre!("path for test `{}` has no file name", path.display()))?
+        .to_str()
+        .ok_or_else(|| {
+            eyre!(
+                "path for test `{}` is not a valid UTF-8 string",
+                path.display()
+            )
+        })?;
+
+    let metadata = read_metadata(path)?;
+
+    Test::new(name, path, metadata)
+}
+
+/// Reads the metadata from the input test code.
+fn read_metadata(test: &Path) -> Result<MetaData> {
+    use regex::bytes::Regex;
+
+    /// Regular expression to retrieve the metadata of a test.
+    static META_REGEX: OnceLock<Regex> = OnceLock::new();
+
+    let code = fs::read_to_string(test)?;
+
+    let yaml = META_REGEX
+        .get_or_init(|| {
+            Regex::new(r"/\*\-{3}((?:.|\n)*)\-{3}\*/")
+                .expect("could not compile metadata regular expression")
+        })
+        .captures(code.as_bytes())
+        .ok_or_eyre("missing metadata for test")?
+        .get(1)
+        .map(|m| String::from_utf8_lossy(m.as_bytes()))
+        .ok_or_eyre("invalid metadata for test")?
+        .replace('\r', "\n");
+
+    serde_yaml::from_str(&yaml).map_err(Into::into)
+}
```

Comment:
I think we have a better version on `main`: https://github.com/boa-dev/boa/blob/main/tests/tester/src/read.rs#L218-L231

We may want to extract the tester to it's own crate at some point, it's easier now that we have the test262 running nightly, to avoid having so much duplicate code :sweat_smile: